python/iceberg/api/partition_spec.py

```diff
@@ -117,9 +118,9 @@ def compatible_with(self, other):
 
     def lazy_fields_by_source_id(self):
         if self.fields_by_source_id is None:
-            self.fields_by_source_id = dict()
+            self.fields_by_source_id = defaultdict(list)
             for field in self.fields:
-                self.fields_by_source_id[field.source_id] = field
+                self.fields_by_source_id[field.source_id].append(field)
```

Comment:
was wondering about this change...what prompted it? Apologies if I missed it. I noticed the similar method below it `lazy_fields_bu_source_name` still uses only a dict. 

None of the added tests exercise this change either,

Reply:
I believe this is because there may be multiple hidden partitions derived from the same table column.  I think Ryan Blue introduced this in the java implementation here:
https://github.com/apache/iceberg/commit/649cbdde83693ebda8e8dc6e75857426d25414ec#diff-d1905822d843dea78ebe5404ee9ce885b7adbc46970fbaf931a87ae2758abeb6

Let me double check, I had written this in our internal repo a while back and it's a little foggy now.

Reply:
at the very least, I will add some test coverage, but it may be worthwhile looking at the change more critically in general.